source/Nevermore/IQueryBuilder.cs

```diff
@@ -1,40 +1,85 @@
-using Nevermore.Joins;
+using Nevermore.AST;
 
 namespace Nevermore
 {
-    public interface IQueryBuilder<TRecord> where TRecord : class
+    public interface IOrderedQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+    }
+
+    public interface ITableSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
     {
-        IQueryGenerator QueryGenerator { get; }
+        ITableSourceQueryBuilder<TRecord> View(string viewName);
+        ITableSourceQueryBuilder<TRecord> Table(string tableName);
+        ITableSourceQueryBuilder<TRecord> Alias(string tableAlias); 
+        IQueryBuilder<TRecord> Hint(string tableHint);
+        IAliasedSelectSource AsAliasedSource();
+    }
 
+    public interface ISubquerySourceBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        ISubquerySource AsSource();
+        ISubquerySourceBuilder<TRecord> Alias(string subqueryAlias);
+    }
+
+    public interface IJoinSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        IJoinSourceQueryBuilder<TRecord> On(string leftField, JoinOperand operand, string rightField);
+    }
+
+    public interface IQueryBuilder<TRecord> where TRecord : class
+    {
         IQueryBuilder<TRecord> Where(string whereClause);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object value);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object startValue, object endValue);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, IEnumerable<object> values);
-        
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, UnarySqlOperand operand, Parameter parameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, BinarySqlOperand operand, Parameter startValueParameter, Parameter endValueParameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, ArraySqlOperand operand, IEnumerable<Parameter> parameterNames);
+
         IOrderedQueryBuilder<TRecord> OrderBy(string orderByClause);
         IOrderedQueryBuilder<TRecord> OrderByDescending(string orderByClause);
+        IQueryBuilder<TRecord> IgnoreDefaultOrderBy();
 
-        IQueryBuilder<TRecord> Join(IJoin join);
-        IQueryBuilder<TRecord> Parameter(string name, object value);
-        IQueryBuilder<TRecord> LikeParameter(string name, object value);
-        IQueryBuilder<TRecord> LikePipedParameter(string name, object value);
-        IQueryBuilder<TRecord> View(string viewName);
-        IQueryBuilder<TRecord> Table(string tableName);
-        IQueryBuilder<TRecord> Hint(string tableHint);
-        IQueryBuilder<TRecord> NoLock();
+        IQueryBuilder<TRecord> Column(string name);
+        IQueryBuilder<TRecord> Column(string name, string columnAlias);
+        IQueryBuilder<TRecord> Column(string name, string columnAlias, string tableAlias);
+        IQueryBuilder<TRecord> AllColumns();
```

Comment:
Isn't this the default? What does this method end up being used for?

Reply:
Yes this is the default, but only if you haven't specified anything else. So if you want to select all columns (i.e. `*`) and also something else (eg `Row_Number()`), then you need to explicitly specify that you want `AllColumns`.

Kind of the same as order by: if you want to order by `Foo` then by `Id`, you still need to explicitly add the `.OrderBy("Id")` even though `Id` is the default order by.